docs/source/software/mpi.md

```diff
@@ -1,28 +1,325 @@
+
+:::{code-block} c++
+#include <mpi.h>
+#include <stdio.h>
+
+int main(int argc, char** argv) {
+    MPI_Init(NULL, NULL);
+
+    int world_rank;
+    MPI_Comm_rank(MPI_COMM_WORLD, &world_rank);
+
+    int number;
+    if (world_rank == 0) {
+        number = -1;
+        MPI_Send(&number, 1, MPI_INT, 1, 0, MPI_COMM_WORLD);
+    } else if (world_rank == 1) {
+        MPI_Recv(&number, 1, MPI_INT, 0, 0, MPI_COMM_WORLD, MPI_STATUS_IGNORE);
+        printf("Process 1 received number %d from process 0\n", number);
+    }
+
+    MPI_Finalize();
+}
+:::
+
+### MPI Monte Carlo
+
+A key aspect of using OpenMPI is the ability to implement parallel algorithms, which can significantly speed up computation. Here is an example of a parallel version of the Monte Carlo method for estimating the number π:
+
+:::{code-block} c++
+#include <mpi.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <time.h>
+
+int main(int argc, char** argv) {
+    MPI_Init(NULL, NULL);
+
+    int world_rank;
+    MPI_Comm_rank(MPI_COMM_WORLD, &world_rank);
+    srand(time(NULL) * world_rank); // Ensure random numbers on all processes
+
+    int local_count = 0;
+    int global_count = 0;
+    int flip = 1 << 24;
+    double x, y, z;
+
+    // Calculate hits within circle locally
+    for (int i = 0; i < flip; i++) {
+        x = (double)rand() / (double)RAND_MAX;
+        y = (double)rand() / (double)RAND_MAX;
+        z = sqrt((x*x) + (y*y));
+        if (z <= 1.0) {
+            local_count++;
+        }
+    }
+
+    // Combine all local sums into the global sum
+    MPI_Reduce(&local_count, &global_count, 1, MPI_INT, MPI_SUM, 0, MPI_COMM_WORLD);
+
+    // Process 0 calculates pi and prints the result
+    if (world_rank == 0) {
+        double pi = ((double)global_count / (double)(flip * world_rank)) * 4.0;
+        printf("The estimated value of pi is %f\n", pi);
+    }
+
+    MPI_Finalize();
+}
+:::
+
+In this code, each process performs its own Monte Carlo simulation and then combines its results with those from other processes using the `MPI_Reduce` function.
+
+### Using OpenMPI with Python's mpi4py
+
+mpi4py is a Python package that provides bindings to the MPI standard. It allows Python programs to take advantage of the distributed memory model and scale across multiple nodes of a high performance computing cluster, just like MPI.
+
+The mpi4py package has been designed to be as close as possible to the MPI standard, providing Python developers with a familiar and straightforward interface to MPI.
+
+In this program, process 0 sends the number -1 to process 1, which receives it and prints it.
+
+To install mpi4py inside of a conda environment:
+
+:::{code-block} bash
+srun -n 4 --pty /bin/bash
+module load anaconda3/2022.05
+source activate <mpi4py_env>
```

Comment:
Are you missing the conda create command here?

Reply:
Good catch. I corrected that and pointed to using the prefix command as we do in the conda documentation. Thank you!!